apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -77,6 +66,204 @@ mod prepare_block {
+
+            let vote_extension_digest =
+                local_last_commit.and_then(|local_last_commit| {
+                    let votes = local_last_commit.votes;
+                    self.compress_vote_extensions(votes)
+                });
+            let vote_extension_digest =
+                match (vote_extension_digest, self.storage.last_height) {
+                    // handle genesis block
+                    (None, BlockHeight(0)) => return vec![],
+                    (Some(_), BlockHeight(0)) => {
+                        tracing::error!(
+                            "The genesis block should not contain vote \
+                             extensions"
+                        );
+                        // TODO: maybe slash validators who claim to have
+                        // seen vote extensions at H=0
+                        return vec![];
+                    }
+                    // handle block heights > 0
+                    (Some(digest), _) => digest,
+                    _ => unreachable!(
+                        "Honest Namada validators will always sign a \
+                         VoteExtension, even if no Ethereum events were \
+                         observed at a given block height. In fact, signing \
+                         an empty VoteExtension commits the fact no events \
+                         were observed by a majority of validators. Likewise, \
+                         a Tendermint quorum should never decide on a block \
+                         including vote extensions reflecting less than or \
+                         equal to 2/3 of the total stake. These scenarios are \
+                         virtually impossible, so we will panic here."
+                    ),
+                };
+
+            let tx = ProtocolTxType::EthereumEvents(vote_extension_digest)
+                .sign(&protocol_key)
+                .to_bytes();
+            let tx_record = record::add(tx);
+
+            vec![tx_record]
+        }
+
+        /// Builds a batch of mempool transactions
+        fn build_mempool_txs(&mut self, txs: Vec<Vec<u8>>) -> Vec<TxRecord> {
+            // filter in half of the new txs from Tendermint, only keeping
+            // wrappers
+            let number_of_new_txs = 1 + txs.len() / 2;
+            txs.into_iter()
+                .take(number_of_new_txs)
+                .map(|tx_bytes| {
+                    if let Ok(Ok(TxType::Wrapper(_))) =
+                        Tx::try_from(tx_bytes.as_slice()).map(process_tx)
+                    {
+                        record::keep(tx_bytes)
+                    } else {
+                        record::remove(tx_bytes)
+                    }
+                })
+                .collect()
+        }
+
+        /// Builds a batch of DKG decrypted transactions
+        fn build_decrypted_txs(&mut self) -> Vec<TxRecord> {
+            // TODO: This should not be hardcoded
```

Comment:
This isn't related to the PR but since we're refactoring here. I think this TODO should be rewritten and maybe link to an issue explaining what the status is with the DKG/front running protection. As I understand for Namada, this encryption step doesn't provide any real protection to transactions right now but Namada specifically won't have DKG/front running protection as a feature.